### PR TITLE
Patch Leaflet to be Server-Side-Render friendly (SSR)

### PR DESCRIPTION
**Pull request by matthova**
This Pull request adds the eslint plugin `ssr-friendly` and implements changes to meet the new linting requirements.

I've confirmed I'm able to use the forked library with a SSR Remix app.

### CODE REVIEW

File: docs/docs/js/docs.js [hunk 1]
```diff
@@ -2,85 +2,92 @@
 hljs.configure({tabReplace: '    '});
 hljs.initHighlighting();
 
-const tocCopy = document.createElement('div');
-tocCopy.id = 'toc-copy';
+function setup() {
```
Comment (IvanSanchez) [inline docs/docs/js/docs.js:5]: Anything in `docs/` will only ever run in a browser. You made the blanket assumption that all `.js` files are part of the Leaflet code, but some are part of its website. These changes are therefore too broad.

File: src/core/Browser.js [hunk 2]
```diff
@@ -13,54 +13,71 @@
...
-	safari,
-	mobile,
-	pointer,
-	touch,
-	touchNative,
-	retina,
-	mac,
-	linux
-};
+export default setupBrowser();
```
Comment (IvanSanchez) [inline src/core/Browser.js:83]: This breaks tree-shaking (which, from my PoV, is more important than SSR friendlyness)

File: src/core/Browser.js [hunk 3]
```diff
@@ -13,54 +13,71 @@
...
-const mobile = typeof orientation !== 'undefined' || userAgentContains('mobile');
+	// @property mobile: Boolean; `true` for all browsers running in a mobile device.
+	const mobile = typeof orientation !== 'undefined' || userAgentContains('mobile');
 
-// @property pointer: Boolean
-// `true` for all browsers supporting [pointer events](https://msdn.microsoft.com/en-us/library/dn433244%28v=vs.85%29.aspx).
-const pointer = !!window.PointerEvent;
+	// @property pointer: Boolean
+	// `true` for all browsers supporting [pointer events](https://msdn.microsoft.com/en-us/library/dn433244%28v=vs.85%29.aspx).
+	const pointer = !!window.PointerEvent;
```
Comment (IvanSanchez) [inline src/core/Browser.js:41]: You may use something like `!!(window?.PointerEvent)`. Ditto for `navigator` and the like.

**Review by IvanSanchez — CHANGES_REQUESTED** [review]

**Comment by IvanSanchez** [comment]
And do read https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining .

File: src/core/Browser.js (hunk 2 above)
Comment (matthova) [reply src/core/Browser.js:83]: Good callout. I'll see if there's a way to have both

**Comment by matthova** [comment]
@IvanSanchez thank you for the feedback! I love the idea of enforcing ssr-compatibility via the lint plugin I added, however, I wonder if there's a slightly different approach we could take for this library to remove unnecessary function-wrapping. Let me do some digging

**Comment by jonkoops** [comment]
> I've confirmed I'm able to use the forked library with a SSR Remix app.

I am curious about this. Why would Leaflet need to be 'SSR friendly' in order for this to work? Leaflet was always intended to be a complete client-side solution, so why even use it in an SSR context? Can Remix not opt-out of SSR for the few places one might need to include Leaflet?

**Comment by matthova** [comment]
> > I've confirmed I'm able to use the forked library with a SSR Remix app.
> 
> I am curious about this. Why would Leaflet need to be 'SSR friendly' in order for this to work? Leaflet was always intended to be a complete client-side solution, so why even use it in an SSR context? Can Remix not opt-out of SSR for the few places one might need to include Leaflet?

That's a great question. I think the maintainers of the library need to own this decision. My use case, as a consumer, is that I love being able to install a package and not have to think about it or do any additional configuration. The tradeoff may not be worth it, for maintainers.

**Comment by matthova** [comment]
Closing and will submit an alternate approach with optional chaining

**Comment by matthova** [comment]
Alternate PR: https://github.com/Leaflet/Leaflet/pull/9317

**Pull request closed without merging**